File: matlab_codes/volume_view.m

Nx = 384-12; Ny = 512-12; Nz = 512-12;

vol = uint8(ones(Nx,Ny,Nz));

thickness = 12;      % Diameter of the cylinder
radius = thickness / 2;

% 5 uniformly spaced X positions (This defines our 5 columns)
xpos = round(linspace(30, Nx-thickness-30, 5));

% --- Calculate Y positions for staggered columns ---
% 1. Positions for the 5-rod columns
ypos5 = round(linspace(30, Ny-thickness-30, 5));

% 2. Calculate the distance between adjacent rods in the 5-rod column
dy = ypos5(2) - ypos5(1);

% 3. Positions for the 4-rod columns (shifted down by exactly half the distance)
ypos4 = round(ypos5(1:4) + dy/2);

% --- Create a 3D Cylindrical Mask ---
% 1. Create a 2D local grid for the cross-section (in X-Y plane)
[xx, yy] = ndgrid(1:thickness, 1:thickness);
xc = radius + 0.5; % X center
yc = radius + 0.5; % Y center

% 2. Create the 2D circular mask using the circle equation: (x-xc)^2 + (y-yc)^2 <= r^2
circle_mask_2D = ((xx - xc).^2 + (yy - yc).^2) <= radius^2;

% 3. Expand the 2D mask into a 3D cylinder that stretches across the entire Z axis
circle_mask_3D = reshape(circle_mask_2D, [thickness, thickness, 1]);
circle_mask_3D = repmat(circle_mask_3D, [1, 1, Nz]);
% -------------------------------------

%--------------------------------------------------
% 23 rods total (5+4+5+4+5) -> Circular Cylinders
% Rods run along Z
% Staggered layout along Y, Zig-zag ordering in Y
%--------------------------------------------------

for kx = 1:5
    
    x = xpos(kx);
    
    % Determine if this column has 5 rods (odd kx) or 4 rods (even kx)
    if mod(kx,2) == 1
        num_rods = 5;
        current_ypos = ypos5;
        yorder = 1:num_rods;      % top -> bottom
    else
        num_rods = 4;
        current_ypos = ypos4;
        yorder = num_rods:-1:1;   % bottom -> top
    end
    
    for ky = 1:num_rods
        
        % Pick the Y position from the correct array based on the column type
        y = current_ypos(yorder(ky));
        
        % Define the bounding box indices for this specific rod
        x_idx = x : x+thickness-1;
        y_idx = y : y+thickness-1;
        
        % Extract the current background subvolume (grabbing all Z slices)
        sub_vol = vol(x_idx, y_idx, :);
        
        % Apply the blood material (2) only where the cylindrical mask is true
        sub_vol(circle_mask_3D) = 2;
        
        % Place the updated subvolume back into the main volume
        vol(x_idx, y_idx, :) = sub_vol;
        
    end
end


cfg.vol = vol;

% ========================================================
% --- 4. 3D VISUALIZATION ---
% ========================================================

figure('Color', 'white', 'Name', '3D Volume Viewer');
hold on;

% A. Extract the surface of the blood rods (where vol == 2)
% We use 1.5 as the threshold boundary between 1 (background) and 2 (blood)
disp('Extracting 3D geometry... this may take a few seconds.');
[faces, vertices] = isosurface(vol == 2, 0.5); 

% B. Render the rods in BLUE
p = patch('Faces', faces, 'Vertices', vertices);
p.FaceColor = '#0055A4';    % A nice, deep professional blue
p.EdgeColor = 'none';       % Remove the mesh lines for smooth rendering

% C. Add lighting and shading to make the rods look 3D and cylindrical
camlight; 
lighting gouraud;
material shiny;             % Adds a slight glossy highlight

% D. Draw the RED REFERENCE BOUNDARY at Z = 250
% This draws a rectangle around the outer edge (1 to Nx, 1 to Ny) at the Z=250 height
Z_ref = 250;
x_rect = [1, Nx, Nx, 1, 1];
y_rect = [1, 1, Ny, Ny, 1];
z_rect = [Z_ref, Z_ref, Z_ref, Z_ref, Z_ref];

plot3(x_rect, y_rect, z_rect, 'Color', 'red', 'LineWidth', 3);

% E. Format the View and Axes
view(3);                   % Set to default 3D isometric view
axis equal;                % Ensure the aspect ratio is correct (1 voxel = 1 unit)
grid on;
box on;                    % Show the outer bounding box of the whole space

% Set axis limits tightly around the volume
xlim([1 Nx]);
ylim([1 Ny]);
zlim([1 Nz]);

xlabel('X-axis');
ylabel('Y-axis');
zlabel('Z-axis');
title('Blue Rods with Red Z=250 Reference Boundary');

hold off;
disp('Rendering complete.');

File: python_codes/volume_view.py
# import numpy as np

# # --- 1. Define Dimensions ---
# Nx = 384 - 12
# Ny = 512 - 12
# Nz = 512 - 12

# # Initialize the volume with 1s (Background). uint8 saves memory.
# vol = np.ones((Nx, Ny, Nz), dtype=np.uint8)

# thickness = 12       # Diameter of the cylinder
# radius = thickness / 2.0

# # 5 uniformly spaced positions 
# # (subtracting 1 to convert MATLAB's 1-based to Python's 0-based index)
# xpos = np.round(np.linspace(40, Nx - thickness - 40, 5)).astype(int) - 1
# ypos = np.round(np.linspace(30, Ny - thickness - 30, 5)).astype(int) - 1

# # --- 2. Create a 3D Cylindrical Mask ---
# # Create a 2D local grid (indexing='ij' matches MATLAB's ndgrid)
# xx, yy = np.meshgrid(np.arange(1, thickness + 1), 
#                      np.arange(1, thickness + 1), 
#                      indexing='ij')
# xc = radius + 0.5
# yc = radius + 0.5

# # 2D circular mask using the circle equation
# circle_mask_2D = ((xx - xc)**2 + (yy - yc)**2) <= radius**2

# # Expand the 2D mask into a 3D cylinder that stretches across the Z axis
# # Add a new axis and repeat it Nz times
# circle_mask_3D = np.repeat(circle_mask_2D[:, :, np.newaxis], Nz, axis=2)

# # --- 3. Place 25 Rods (5x5) ---
# for ky in range(5):
#     # Zig-zag ordering in X as Y increases
#     # In Python, ky=0,2,4 are the "odds" if we compare to MATLAB's 1,3,5
#     if ky % 2 == 0:
#         xorder = list(range(5))          # left -> right
#     else:
#         xorder = list(range(4, -1, -1))  # right -> left
        
#     y = ypos[ky]
    
#     for kx in range(5):
#         x = xpos[xorder[kx]]
        
#         # Extract the current background subvolume (all Z slices)
#         sub_vol = vol[x:x+thickness, y:y+thickness, :]
        
#         # Apply the blood material (2) using boolean masking
#         sub_vol[circle_mask_3D] = 2
        
#         # Place updated subvolume back into the main volume
#         vol[x:x+thickness, y:y+thickness, :] = sub_vol

# # Now 'vol' contains your 3D array

# import pyvista as pv

# # 1. Map the NumPy array to a PyVista spatial grid
# grid = pv.ImageData(dimensions=vol.shape)

# # Flatten in Fortran order ('F') to match PyVista/VTK memory layout
# grid.point_data["Materials"] = vol.flatten(order="F")

# # 2. Extract ONLY the blood rods (where value == 2)
# # We use a threshold to hide the background (1s)
# rods = grid.threshold([1.5, 2.5])

# # 3. Setup the Plotter (Publication Quality)
# plotter = pv.Plotter()
# plotter.background_color = "white"

# # Add the rods to the plot
# plotter.add_mesh(rods, color="red", smooth_shading=True, label="Blood Vessels")

# # Add the "3D Box" (bounding box) around the entire volume space
# plotter.add_mesh(grid.outline(), color="black", line_width=3)

# # Add clear axes and a scale
# plotter.show_axes()
# plotter.add_bounding_box(color="gray", line_width=1)

# # 4. Display and Save
# # Uncomment the line below to save a high-res image for your paper. 
# # scale=3 multiplies the window resolution by 3.
# # plotter.screenshot("volume_render_high_res.png", scale=3) 

# plotter.show()

# # ==========================================================================
# # ==========================================================================


# import numpy as np
# import matplotlib.pyplot as plt
# from skimage import measure
# from itertools import product, combinations

# # --- Define Volume ---
# Nx, Ny, Nz = 384-12, 512-12, 512-12
# vol = np.ones((Nx, Ny, Nz), dtype=np.uint8)

# thickness = 12
# radius = thickness / 2.0

# xpos = np.round(np.linspace(40, Nx - thickness - 40, 5)).astype(int)
# ypos = np.round(np.linspace(30, Ny - thickness - 30, 5)).astype(int)

# # --- Create a 3D Cylindrical Mask ---
# xx, yy = np.mgrid[1:thickness+1, 1:thickness+1]
# xc, yc = radius + 0.5, radius + 0.5
# circle_mask_2D = ((xx - xc)**2 + (yy - yc)**2) <= radius**2
# circle_mask_3D = np.repeat(circle_mask_2D[:, :, np.newaxis], Nz, axis=2)

# # --- Place 25 Rods (5x5) into the Volume ---
# for ky in range(5):
#     if ky % 2 == 0:
#         xorder = range(5)          
#     else:
#         xorder = range(4, -1, -1)  
        
#     y = ypos[ky] - 1
    
#     for kx in range(5):
#         x = xpos[xorder[kx]] - 1
#         sub_vol = vol[x:x+thickness, y:y+thickness, :]
#         sub_vol[circle_mask_3D] = 2

# # ==========================================
# # 3D Visualization (Clean View)
# # ==========================================
# print("Extracting surface for 3D rendering (this may take a moment)...")

# step = 4
# vol_ds = vol[::step, ::step, ::step]

# # Marching Cubes
# verts, faces, normals, values = measure.marching_cubes(vol_ds, level=1.5)
# verts = verts * step  

# # Initialize plot
# fig = plt.figure(figsize=(10, 8))
# ax = fig.add_subplot(111, projection='3d')

# # Plot the rods in green
# ax.plot_trisurf(verts[:, 0], verts[:, 1], faces, verts[:, 2], 
#                 color='green', alpha=0.9, edgecolor='none')

# # Draw the black boundary box
# r_x = [0, Nx]
# r_y = [0, Ny]
# r_z = [0, Nz]

# for s, e in combinations(np.array(list(product(r_x, r_y, r_z))), 2):
#     dist = np.sum(np.abs(s - e))
#     if dist == Nx or dist == Ny or dist == Nz:
#         ax.plot3D(*zip(s, e), color="black", linewidth=1.5)

# # Keep proportional aspect ratio
# ax.set_box_aspect([Nx, Ny, Nz])  

# # --- Remove Axes, Ticks, and Grids ---
# ax.set_axis_off() 

# # Remove extra padding
# plt.tight_layout(pad=0)
# plt.show()



# # ==========================================================================
# # ==========================================================================


import numpy as np
import pyvista as pv

# --- 1. Define Volume ---
Nx, Ny, Nz = 384-12, 512-12, 512-12
vol = np.ones((Nx, Ny, Nz), dtype=np.uint8)

thickness = 12
radius = thickness / 2.0

xpos = np.round(np.linspace(40, Nx - thickness - 40, 5)).astype(int)
ypos = np.round(np.linspace(30, Ny - thickness - 30, 5)).astype(int)

# --- 2. Create Mask ---
xx, yy = np.mgrid[1:thickness+1, 1:thickness+1]
xc, yc = radius + 0.5, radius + 0.5
circle_mask_2D = ((xx - xc)**2 + (yy - yc)**2) <= radius**2
circle_mask_3D = np.repeat(circle_mask_2D[:, :, np.newaxis], Nz, axis=2)

# --- 3. Place Rods ---
for ky in range(5):
    if ky % 2 == 0:
        xorder = range(5)          
    else:
        xorder = range(4, -1, -1)  
        
    y = ypos[ky] - 1
    
    for kx in range(5):
        x = xpos[xorder[kx]] - 1
        sub_vol = vol[x:x+thickness, y:y+thickness, :]
        sub_vol[circle_mask_3D] = 2

# ==========================================
# 3D Visualization using PyVista (FIXED)
# ==========================================
# ==========================================
# 3D Visualization using PyVista (FIXED + SLICE)
# ==========================================
print("Generating 3D render...")

# Wrap the numpy array into a PyVista grid
grid = pv.ImageData()
grid.dimensions = vol.shape 
grid.spacing = (1, 1, 1)
grid.point_data["values"] = vol.flatten(order="F") 

# Extract the surface where value == 2
mesh = grid.contour([1.5], scalars="values")

# Set up the plotter
plotter = pv.Plotter()

# 1. Add the rods (Green, smooth shading)
plotter.add_mesh(mesh, color='green', smooth_shading=True)

# 2. Add a bounding box outline (Black)
plotter.add_mesh(grid.outline(), color='black', line_width=2)

# 3. --- NEW: Highlight Z = 250 slice ---
z_slice = 250
# Create a plane centered in X and Y, sitting at Z = 250
slice_plane = pv.Plane(center=(Nx/2, Ny/2, z_slice), 
                       direction=(0, 0, 1), # Normal vector pointing up Z
                       i_size=Nx, j_size=Ny)

# Plot the plane with a thick blue border and slight blue tint
plotter.add_mesh(slice_plane, 
                 color='blue', 
                 opacity=0.2,       # Slight transparent fill to show the cut
                 show_edges=True,   # Turn on the boundary line
                 edge_color='blue', # Make boundary blue
                 line_width=5)      # Make the boundary thick and visible

# Remove the background grid/axes for a clean view
plotter.set_background('white')
plotter.hide_axes()

# Show the interactive window
plotter.show()
